File: common/src/main/java/dev/obscuria/fragmentum/v2/core/common/registry/_Deferred.java
package dev.obscuria.fragmentum.v2.core.common.registry;

import dev.obscuria.fragmentum.v2.api.common.registry.Deferred;
import net.minecraft.core.Holder;

public class _Deferred<T> implements Deferred<T> {

    protected final Holder<T> holder;

    public _Deferred(Holder<T> holder) {
        this.holder = holder;
    }

    @Override public Holder<T> holder() {
        return holder;
    }

    @Override public T get() {
        return holder().value();
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/api/registry/Deferred.kt
package dev.obscuria.fragmentum.api.registry

import net.minecraft.core.Holder

@Suppress("UNCHECKED_CAST")
open class Deferred<T, V : T>(supplier: () -> Holder<T>) : HolderProvider<T>, ValueProvider<V> {

    override val holder: Holder<T> by lazy { supplier.invoke() }
    override val value: V by lazy { holder.value() as V }
}